lib/cc/workspace/path_tree.rb

```diff
@@ -0,0 +1,115 @@
+require "pathname"
+require "set"
+
+module CC
+  class Workspace
+    class PathTree
+      def initialize(root_path, children = {})
+        @root_path = root_path.dup.freeze
+        @children = children
+      end
+
+      def clone
+        self.class.new(root_path, children.dup)
+      end
+
+      def exclude_paths(paths)
+        paths.each { |path| delete(normalized_path_pieces(path)) }
+      end
+
+      def include_paths(paths)
+        paths.each { |path| insert(normalized_path_pieces(path)) }
+      end
+
+      def all_paths
+        if populated?
+          children.values.flat_map(&:all_paths)
+        else
+          [File.join(root_path, File::SEPARATOR)]
+        end
+      end
+
+      protected
+
+      def populated?
+        children.present?
+      end
+
+      def delete(path_pieces)
```

Comment:
FWIW: If you made this method signature `delete(head, *tail)`, you'd get `path_pieces[0]` and `path_pieces.drop(1)` for free. This is a common way to operate on lists in Haskell.